enable_disable_persistence.md

```diff
@@ -0,0 +1,195 @@
+single-click enable/disable interaction to work.
+
+For more details, see [Enable/Disable Resource API Spec](./enable_disable_resources.md).
+
+But we have different ideas on how we want the enable/disable interaction to work
+across multiple Tilt runs, and how the state should persist! [^1]
+
+## Primary Goal
+
+The goal of this doc is to give a product specification by example: describe
+specific sequences of user actions, and how we want Tilt to behave. (Note that
+it is not intended as a description of how Tilt behaves today.)
+
+## Non-Goals
+
+### Groups
+
+We expect that we will eventually want to control groups of services, but
+we don't believe that changes our analysis here about how single services behave.
+
+## Terminology
+
+### Enabled
+
+We say a server is "enabled" if it's visible in the Tilt UI and is in status
+waiting on trigger, pending on dependency, running, completed, or error.
+
+### Disabled
+
+We say a server is "disabled" if it's visible in the Tilt UI and is in status
+"disabled". We do not expect any new logs or status changes in the UI until the
+service is intentionally "enabled"
+
+### Session
+
+We say a Tilt "session" is the time from when the user runs "tilt up" to when
+the tilt process exits.
+
+### Service Names
+
+For readability, we'll use NY pizza places as service names (Paulie, Juliana, Roberta, Joe).
+
+## Prior Art
+
+### Resource Pinning
+
+Today, when you pin a resource in the Tilt UI, the pin persists across Tilt sessions.
+
+The state is stored the browser's LocalStore, and is tied to the user's browser.
+
+### Manual Trigger
+
+Today, when you switch a resource between manual trigger mode and auto trigger
+mode, that state persists until you reload the Tiltfile. When the Tiltfile
+loads/reloads, the trigger mode reverts to what's specified in the Tiltfile.
+
+### systemd
+
+[systemd and
+systemctl](https://www.commandlinux.com/man-page/man1/systemctl.1.html) is
+probably the most widely known service manager. You tend to see its verbs in
+other service management tools (like
+[Edward](http://engblog.yext.com/edward/commands/#start)).
+
+`systemd` is a persistent daemon that runs admin-level services on linux.
+
+You use `systemctl start` and `systemctl stop` to start/stop services.
+
+And `systemctl enable` and `systemctl disable` to enable/disable services to
+auto-boot at system startup.
+
+Note that `enable` does not start a service immediately. In other words, systemd
+models "start on boot" and "start now" as independent actions.
+
+## Product Specification
+
+### Run all, Disable from UI
+
+1. Write a Tiltfile with services Paulie and Roberta
+1. Run `tilt up` -> Paulie and Roberta are enabled.
+1. Click "disable" on "Paulie" -> Roberta is enabled, Paulie is disabled
+1. Ctrl-c and re-run `tilt up` -> Roberta is enabled, Paulie is disabled
+
+Possible Principle:
+
+`tilt up` (without arguments) always inherits all the enable/disable overides
+from the previous Tilt session.
```

Comment:
i could see this perhaps being unexpected or confusing behavior as a principle. (but i could be wrong!) it could be nice functionality to have as an opt-in flag, like `tilt up --use-last-run`.

for me, i think there is something intuitive in the difference between:
- tiltfile: these are my saved settings
- tilt up: this is how i'm configuring this session, but don't persist across sessions
- dashboard/cli: these settings are changed within a session, but don't persist across sessions

Reply:
L owns this decision, and is responsible for collecting feedback on whether it's the right one. If we get new info on this, you should raise that with them. 

The reason I wrote this doc is because they said, "hey Nick, I've decided that 'tilt up' should persist enable/disable state, can you write up a doc detailing all the implications of that?" Which is here! 

I have my own product opinions on this, but let's put them aside. If I was putting my product critique hat on your proposal, I'd raise an eyebrow at this statement:

> tilt up: this is how i'm configuring this session

If that's your user story, then it seems like you'd want the configuration parameters to be mandatory? i.e., 'tilt up' with no arguments would be an error because no configuration is provided.

Reply:
Lizz, I disagree with `tilt up: [...] don't persist across sessions` because in my experience that's not how most dev apps behave nowadays. To take VSCode as an example:

I get VSCode up with `code -n ~/code/project`, and open a bunch of files in it. At the end of the day, I turn off the computer. Monday I get back to work, click the VSCode icon, and everything's how I left it. Maybe I decide to work on a different project so I run `code -r ~/code/different` which replaces whatever I had open there with the new project. Once I'm done with that, I then Ctrl+Shift+O and load up let's say `~/code/another` via VSCode's UI. I close VSCode, and when I open it again tomorrow morning, it shows me `~/code/another`, the last place I was at.

So I'd rather follow convention and have Tilt mimic that. It might not be possible since there's a lot of extra complexity, but I think at least `tilt up` with no arguments should.

A simpler solution which might just end the whole discussion could be:

```
$ tilt up
Continue where you left off [Y/n]?
```

`Y` shows the user the exact state of the last session, regardless of config file changes. `N` brings up a clean slate, regardless of what the previous state was. `tilt up` with extra arguments behaves as Nick described.

Whether this prompt appears or not, whether it's `[Y/n]` or `[y/N]`, and whether it's `Y` or `N` by default without prompting, can be configured via a `persist_previous_state()` setting on the Tiltfile.

Thoughts?

Reply:
I love this idea - allow the user to decide. Han and I were working off a similar concept.. here's a sneak preview of that:

![Screen Shot 2021-10-01 at 1 58 31 PM](https://user-images.githubusercontent.com/46793797/135666099-e0a282ca-84f9-469b-8369-1b02f9f7cfe0.png)

Reply:
Another idea, from Nick: Always restart from where the user left off, and show a warning that tells the user how to have a clean slate instead if they wish.